docs(*): new policy matching proposal

lobkovilya opened the pull request:
### Summary

[Rendered version to read](https://github.com/kumahq/kuma/blob/docs/target-ref-proposal/docs/madr/decisions/005-policy-matching.md)

### Full changelog

* add a new policy matching proposal

### Issues resolved

Updates https://github.com/kumahq/kuma/issues/3330

### Documentation

- [ ] Link to the website [documentation PR](https://github.com/kumahq/kuma-website/pull/XXX)

### Testing

- [ ] Unit tests
- [ ] E2E tests
- [ ] Manual testing on Universal
- [ ] Manual testing on Kubernetes

### Backwards compatibility

- [ ] Update [`UPGRADE.md`](/UPGRADE.md) with any steps users will need to take when upgrading.
- [ ] Add `backport-to-stable` label if the code follows our [backporting policy](/CONTRIBUTING.md#backporting)

lahabana reviewed (commented):
Great start! I've left a few comments and I think we're missing:

- Showing an example of a policy that's neither inbound nor outbound (did we have one in the end? If not thinking about one to prove our system works would be good).
- Showing an example of a policy that's both inbound and outbound (trafficLog I believe is one of them)
- Schema of TargetRef
- Mention how we're going to roll this out (new policies that will replace existing policies)
- Explain that follow MADR will come to describe each of the new policy
- Explain that the top targetRef always identifies the set of DPP that is being modified

lobkovilya commented:
> Also didn't we talked about a ServiceGroup or ServiceSubset which is like a ProxyGroup except that kuma.io/service is required?

Do you have an example of a policy that supports `ServiceSubset` but doesn't support `ProxyGroup` (or vice versa)?

lobkovilya commented on docs/madr/decisions/005-policy-matching.md:
````diff
@@ -0,0 +1,636 @@
+```yaml
+param1: value1 # overrides 'value2' from 'targetRef{kind:Service,name:backend}' 
+param2: value3 # overrides 'value4' from 'targetRef{kind:Mesh}'
+```
````
Do you mean no difference in merging between `to` and `from`?

lobkovilya replied:
The difference is what is a final consumer of the configuration. It's easy for outbound policies because the final consumer is an outbound (a real object in Envoy proxy). It's less straightforward for inbound policies because the final consumer is a source (an artificial object, different sources represent different clients).

lobkovilya commented:
@lahabana I renamed ProxyGroup to ServiceSubset and added a new MeshSubset level, so now it looks like:
* Mesh
* MeshSubset
* Service
* ServiceSubset
* Proxy
* MeshGatewayRoute 
* HTTPRoute

> Related but not exactly the same what replaces our current `service: *`?

yes, so now if you want `service: *` you'll create:
```yaml
targetRef:
  kind: MeshSubset
  tags:
    key1: value1
```

lahabana commented on docs/madr/decisions/005-policy-matching.md:
```diff
@@ -0,0 +1,647 @@
+### Overview
+
+"TargetRef" selector is inspired by [Gateway API Policy Attachment](https://gateway-api.sigs.k8s.io/v1alpha2/references/policy-attachment/) specification. 
+Even though, Kuma policy matching cannot be compatible with Gateway API Policy Attachment spec at that moment,
```
Probably should say why it cannot be compatible

lahabana commented on docs/madr/decisions/005-policy-matching.md:
```diff
@@ -0,0 +1,647 @@
+message TargetRef {
+  // Kind of the referenced resource
+  // +kubebuilder:validation:Enum=Mesh;MeshSubset;Service;ServiceSubset;Proxy;MeshGatewayRoute;HTTPRoute
+  string kind = 1;
```
Shouldn't there be an `apiVersion` with a good default?

lobkovilya replied:
> Should we remove ambiguity and name Service MeshService ?

Do I understand it right, you want to have something like:
* Mesh
* MeshSubset
* MeshService
* MeshServiceSubset
* Proxy
* MeshGatewayRoute
* HTTPRoute

> HTTPRoute is this just GatewayAPI HTTPRoute (or at least a mirror of it)?

Yes, but without `backends` and with `targetRef` instead of `selectors`. I can add an example of HTTPRoute

lahabana commented on docs/madr/decisions/005-policy-matching.md:
```diff
@@ -0,0 +1,647 @@
+      - targetRef:
+          kind: Service
+          name: payments
+        connectTimeout: 10s
```
while the previous nesting wasn't required. Adding one here might help so that the backbone of the policies is always the same:

```
spec:
  targetRef: {}
  config: {} # Only actual type that differs.
  from:
      - targetRef: {}
        config: {} # Only actual type that differs.
  to:
      - targetRef: {}
        config: {} # Only actual type that differs.
```

lobkovilya replied:
so `spec.config` will be used only for policies that have neither `from` nor `to`, like ProxyTemplate?

lobkovilya replied:
I added `conf` instead of `config` inside each array's item. Top-level `conf` will be used for example for ProxyTemplates

lahabana reviewed (commented):
I don't see anywhere an explanation that states that "to" and "from" will have different valid subsets of "targetRef.kind"

lahabana commented:
> Related but not exactly the same what replaces our current service: *?

So service subset will ALWAYS require a serviceName to be set?

Also MeshSubset will never be usable in `to` and `from` in the end no? Sounds like you can only pick subsets of a service in to and from.

michaelbeaumont commented on docs/madr/decisions/005-policy-matching.md:
````diff
@@ -0,0 +1,647 @@
+resulting configuration for "backend" outbound is:
+
+```yaml
+param1: value1 # overrides 'value2' from 'targetRef{kind:Service,name:backend}' 
````
Hmm, this feels backwards. Shouldn't later entries override earlier ones?

lobkovilya replied:
> outboundConfs are merged in the order they're presented in the "to" array

I meant order from top to bottom, I can make it clearer

michaelbeaumont replied:
I mean I would expect to have `param1: value2` then since the `Service` would override `Mesh`.

lobkovilya replied:
No, it's actually simpler. Only order matters at this point. We merge `outboundConfs` without looking at `targetRefs`

michaelbeaumont commented on docs/madr/decisions/005-policy-matching.md:
````diff
@@ -0,0 +1,647 @@
+    param2: value4
+```
+
+resulting configuration for "backend" outbound is:
````
This way policies are added together here seems like something we might want to be policy specific, should we leave it undefined?

lobkovilya replied:
In the big list of examples, I shared earlier, a simple merge strategy (lists are replaced not appended) worked well for all cases. Do you have anything specific in mind?

michaelbeaumont replied:
Not at the moment, it just doesn't seem necessary to make one rule for all policies.

lobkovilya commented on docs/madr/decisions/005-policy-matching.md:
```diff
@@ -0,0 +1,636 @@
+* ProxyGroup
+* Proxy
+* MeshGatewayRoute 
+* HTTPRoute
```
Yes, probably we will have. Can extend `targetRef` in the future

lobkovilya commented on docs/madr/decisions/005-policy-matching.md:
```diff
@@ -0,0 +1,636 @@
+and able to override behavior specified on higher levels.
+
+If multiple policies target the same object, 
+they'll be sorted in a lexicographic order (using policy name). 
```
I feel like conflict resolution based on timestamps may not work well for us. Especially if Kuma syncs policies from Global to Zone, the order is undefined now. The oldest policy from Global can become the newest policy in Zone.

lobkovilya commented on docs/madr/decisions/005-policy-matching.md:
```diff
@@ -0,0 +1,636 @@
+  targetRef:
+    kind: Service
+    name: backend
+  conf:
```
Delete top-level `conf`, added `conf` inside each item of `to, from` arrays

lobkovilya commented:
> I don't see anywhere an explanation that states that "to" and "from" will have different valid subsets of "targetRef.kind"

That's down to policies, in [usecases](https://gist.github.com/lobkovilya/57ae04411fdddedfb15752c9a18d1108) you can find allowed values for each policy

lobkovilya commented:
> So service subset will ALWAYS require a serviceName to be set?

Yes, I think we can avoid having `*` at all, if you don't care about the service name you should use MeshSubset.

> Also MeshSubset will never be usable in to and from in the end no? Sounds like you can only pick subsets of a service in to and from.

You can use `MeshSubset` in `from` array. In the `to` array you can use neither `MeshSubset` nor `ServiceSubset`. 

> Should we therefore disallow using ServiceSubset in top level? I'm unsure but it might be simpler to users

I'm not sure if more exceptions make it simpler. I'd allow both MeshSubset and ServiceSubset in top level, I think it's clear that the latter is more specific

lobkovilya commented:
Awaiting for decision:
1. When it comes to conflict resolution, do we want to follow Gateway API? [link](https://github.com/kumahq/kuma/pull/4474#discussion_r899047742)
  **answer:** yes/no
2. Do we want to rename `kind: Service` to `MeshService`, `KumaService`, etc? [link](https://github.com/kumahq/kuma/pull/4474#discussion_r907348285)
  **answer:** yes rename to [option]/no
3. When merging items inside `to/from` arrays do we want to change the order and have more specific items at the bottom? [link](https://github.com/kumahq/kuma/pull/4474#discussion_r907379083)
  **answer:** yes/no

jakubdyszkiewicz commented on docs/madr/decisions/005-policy-matching.md:
```diff
@@ -0,0 +1,798 @@
+ProxyTemplate is neither inbound nor outbound policy, 
+so it doesn't have "from" or "to" arrays. 
+There are only "imports" and "modifications" arrays
+which override arrays from other levels.
```
I think we should consider appending arrays or introducing mechanism of composing proxy template. Imagine you have Mesh level ProxyTemplate with 5 modifications (because again, they are not composable). Then you want to do something for one service, you need to copy paste those 5 modifications from Mesh level to not break anything.

michaelbeaumont commented on docs/madr/decisions/005-policy-matching.md:
```diff
@@ -0,0 +1,798 @@
+
+"TargetRef" can attach policy to different objects:
+* Mesh
+* MeshSubset
```
One thing to note here, for strict PolicyAttachment compliance, I think there needs to be a `name` field here, even though it's "meaningless".

lobkovilya replied:
Maybe we should have `name` even for Mesh, you just leave it empty as a user, but Kuma will set it automatically

michaelbeaumont commented on docs/madr/decisions/005-policy-matching.md:
```diff
@@ -0,0 +1,798 @@
+* Mesh
+* MeshSubset
+* MeshService
+* MeshServiceSubset
```
I think for strict PolicyAttachment compliance, there "can't" be a `tags` field on `targetRef`.

lobkovilya replied:
Do you mean we should move `tags` somewhere else?

michaelbeaumont reviewed (commented):
Two comments about [PolicyAttachment compliance](https://gateway-api.sigs.k8s.io/v1alpha2/references/policy-attachment/#target-reference-api)

lahabana commented on docs/madr/decisions/005-policy-matching.md:
````diff
@@ -0,0 +1,1048 @@
+```
+
+"TargetRef" can attach policy to different objects:
+* Mesh
````
@lukidzi suggested adding a 1 liner to describe what these are.

bartsmykla reviewed (approved):
lgtm

slonka commented on docs/madr/decisions/005-policy-matching.md:
```diff
@@ -0,0 +1,1052 @@
+Source and destination selectors (S&D) are trying to solve the problem of mapping between policies and connections. 
+But in the real world of Envoy proxies, 
+there is no such thing as "connection". 
+There are "inbounds" and "outbounds". 
```
nitpick: Envoy does not mention "inbounds" or "outbounds" in the [terminology docs](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/intro/terminology) - there are just `Listeners` with [TrafficDirection](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#envoy-v3-api-enum-config-core-v3-trafficdirection) set to `INBOUND` or `OUTBOUND`. I think that the intended audience will figure it out but for clarity sake we could define the terms "inbounds" and "outbounds".

lobkovilya replied:
Yeah, inbounds and outbounds are something that Kuma is using when configuring Envoy. We defined these terms in [doc](https://kuma.io/docs/dev/explore/dpp/#concepts), but it's still under `dev`. Since this is a MADR that is targeted at developers, I think it's fine not to define inbound and outbound in this doc.

slonka commented on docs/madr/decisions/005-policy-matching.md:
```diff
@@ -0,0 +1,1052 @@
+
+When several targetRefs select the same outbound,
+corresponding confs are merged in the order they're presented in the "to" array 
+(bottom ones have more priority over the top ones):
```
I'm not a native speaker but things can **have** "higher priority" and someone / something can **give** "more priority" to things. Maybe native speakers can chip in - @michaelbeaumont @parkanzky 

```suggestion
(bottom ones have higher priority over the top ones):
```

slonka commented on docs/madr/decisions/005-policy-matching.md:
````diff
@@ -0,0 +1,1052 @@
+
+```yaml
+conf:
+  param1: value2 # overrides 'value1' from 'targetRef{kind:Mesh}'
````
value1 is in `targetRef: MeshServiceSubset` not `targetRef{kind:Mesh}` right?

lobkovilya replied:
Good catch, fixed!

slonka commented on docs/madr/decisions/005-policy-matching.md:
````diff
@@ -0,0 +1,1052 @@
+        - name: file
+```
+
+This policy specifies the following behaviour:
````
so if I wanted to log traffic between service web-1 and web-2 on the web1 outbound I'd do:

```yaml
type: TrafficLog
mesh: mesh-1
name: tl-1
targetRef:
  kind: MeshService
  name: web-1
to:
  - targetRef:
      kind: MeshService
      name: web-2
    conf:
      backends:
        - name: file
```

right?

lobkovilya replied:
Yes, this is correct

slonka reviewed (approved):
some nitpicks and questions, nice work overall 👍
